File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_449/java/Solution2.java
package com.hi.dhl.algorithms.leetcode._449.java;

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/4/29
 *     desc  :
 * </pre>
 */

import com.hi.dhl.algorithms.model.ListNode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution2 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        Deque<Integer> stack1 = new ArrayDeque<>();
        Deque<Integer> stack2 = new ArrayDeque<>();

        while (l1 != null) {
            stack1.push(l1.val);
            l1 = l1.next;
        }

        while (l2 != null) {
            stack2.push(l2.val);
            l2 = l2.next;
        }

        int carry = 0;
        ListNode head = null;
        while (!stack1.isEmpty() || !stack2.isEmpty() || carry > 0) {
            int a = !stack1.isEmpty() ? stack1.pop() : 0;
            int b = !stack2.isEmpty() ? stack2.pop() : 0;
            int sum = a + b + carry;
            carry = sum / 10;
            ListNode node = new ListNode(sum % 10);
            node.next = head;
            head = node;
        }
        return head;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_449/kotlin/Solution2.kt
package com.hi.dhl.algorithms.leetcode._449.kotlin

import com.hi.dhl.algorithms.model.ListNode
import java.util.*

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/4/29
 *     desc  :
 * </pre>
 */
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution2 {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val stack1 = ArrayDeque<Int>()
        val stack2 = ArrayDeque<Int>()

        var p1 = l1
        var p2 = l2
        while (p1 != null) {
            stack1.push(p1.`val`)
            p1 = p1.next
        }

        while (p2 != null) {
            stack2.push(p2.`val`)
            p2 = p2.next
        }

        var carry = 0
        var head: ListNode? = null
        while (!stack1.isEmpty() || !stack2.isEmpty() || carry > 0) {
            val a = if (!stack1.isEmpty()) stack1.pop() else 0
            val b = if (!stack2.isEmpty()) stack2.pop() else 0
            val sum = a + b + carry
            carry = sum / 10
            val node = ListNode(sum % 10)
            node.next = head
            head = node
        }
        return head
    }
}